### PR TITLE
Benchmarking

### LINKED ISSUE #78: Benchmarking and Query Tests
Tasks:
Time to simplify the original trajectories with varying parameters

Perform queries on original and simplified tables and check query accuracy

Test whether Knn queries are necessary when determining query accuracy or if range query is only needed:
Knn queries are much slower than range queries, so if we can avoid them in query accuracy that would be great. If we find that simplifications only using range queries are of same size as those using both types of queries, we can drop doing knn. If we find that simplifications using only range queries are much smaller than those using both types of queries, we can infer that knn queries provide value and therefore keep them, sadly.

### PR DESCRIPTION
**Pull request by KarmaKamikaze**
**Summary of added content:**

- Changed Trace-Q to use F1 scores
- Separated min_query_accuracy into 2 values
- Changed knn query test check to not consider ordering
  - This is because the paper we want to compare accuracy against, doesn't consider ordering of results 
- Created Benchmark class that can be used to determine query accuracy on simplified data
  - This closely resembles the functionality of trace_q

**Relevant files for review:**

- `src/benchmark/Benchmark.cpp`
- `src/benchmark/Benchmark.hpp`
- `src/benchmark/F1.cpp`
- `src/benchmark/F1.hpp`
- `src/benchmark/TRACE_Q_Benchmark.cpp`
- `src/benchmark/benchmark-query-objects/Benchmark_KNN_Query.hpp`
- `src/benchmark/benchmark-query-objects/Benchmark_Query.hpp`
- `src/benchmark/benchmark-query-objects/Benchmark_Range_Query.hpp`
- `src/querying/KNN_Query_Test.cpp`
- `src/querying/KNN_Query_Test.hpp`
- `src/simp-algorithms/TRACE_Q.cpp`
- `src/simp-algorithms/TRACE_Q.hpp`
- `src/trajectory_data_handling/Trajectory_Manager.cpp`

### CODE REVIEW

File: src/benchmark/Benchmark.cpp [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+
+                    return F1{knn_query->query_result_ids, simplified_set};
+                }));
+            }
+
+            query_count++;
+
+            if (query_count > max_connections) {
+                for (auto & fut : range_futures) {
+                    range_query_cum_F1.push_back(fut.get());
```
Comment (Flaiizor) [inline src/benchmark/Benchmark.cpp:80]: cum? yeah sure

**Review by Mailaur — APPROVED** [review]

File: src/simp-algorithms/TRACE_Q.cpp [hunk 2]
```diff
@@ -35,61 +35,90 @@ namespace trace_q {
...
-
-                    knn_futures.emplace_back(std::async(std::launch::async, knn_query_initialization, x_coord, y_coord, t_interval, knn_query_mbr, knn_query_time_interval_multiplier, knn_k));
+        if (use_KNN_for_query_accuracy) {
+            auto knn_query_mbr = expand_MBR(calculate_MBR(original_trajectory), knn_query_grid_expansion_factor);
+
+            auto knn_query_points_on_axis = static_cast<int>(std::ceil(1 / knn_query_grid_density));
+            auto knn_query_time_points = static_cast<int>(std::ceil(1 / knn_query_time_interval_multiplier));
+
+            // Here we run the knn queries asynchronously for each query time point.
+            // Note that here, the query time points should not be larger than the allowed connections to the database (100)
```
Comment (Flaiizor) [inline src/simp-algorithms/TRACE_Q.cpp:71]: troede max var 50 og ikke 100?

**Review by Flaiizor — APPROVED** [review]
Don't ever make me read through that amount of text again, unless you write "cum" some more

File: src/simp-algorithms/TRACE_Q.cpp (hunk 2 above)
Comment (KarmaKamikaze) [reply src/simp-algorithms/TRACE_Q.cpp:71]: Det er max, men vi sætter en grænse på 50, da der kan åbnes connections af postgres' egen proces, som vi ikke kan styre. Så det er bare for at være på den sikre side og vi oplever stadig at programmet udnytter 100% af CPUen.

File: src/benchmark/F1.cpp [hunk 3]
```diff
@@ -0,0 +1,48 @@
+#include "F1.hpp"
+
+namespace analytics {
+
+    F1::F1(std::vector<spatial_queries::KNN_Query::KNN_Result_Element> const& original_vec,
+    std::vector<spatial_queries::KNN_Query::KNN_Result_Element> const& simplified_vec) {
+        std::unordered_set<unsigned int> simplified_set{};
+
+        for (const auto& s : simplified_vec) {
```
Comment (Kristoffermg) [inline src/benchmark/F1.cpp]: Der burde stå "simplified" i stedet for s, i for løkken lige under står der "original" i stedet for "o"

**Review by Schwartz123 — APPROVED** [review]
looks good

File: src/simp-algorithms/TRACE_Q.cpp (hunk 2 above)
Comment (Flaiizor) [reply src/simp-algorithms/TRACE_Q.cpp:71]: Very cool

**Review by Kristoffermg — APPROVED** [review]

File: src/benchmark/F1.cpp (hunk 3 above)
Comment (Sarmisuper) [reply src/benchmark/F1.cpp]: Yes, hele den kodeblok bliver dog fjernet, pga at vi laver det om til kun at bruge én constructor.

**Pull request merged**